api/src/main/java/org/eclipse/microprofile/metrics/MetricID.java

```diff
@@ -164,11 +164,26 @@ public String getTagsAsString() {
      * @return a a list of Tag objects
      */
     public List<Tag> getTagsAsList() {
-        List<Tag> list = new ArrayList<Tag>();
+        List<Tag> list = new ArrayList<>();
         this.tags.forEach((key, value) -> list.add(new Tag(key, value)));
         return Collections.unmodifiableList(list);
     }
 
+    /**
+     * Gets the list of tags as an array of {@link Tag} objects.
+     *
+     * @return An array of tags
+     */
+    public Tag[] getTagsAsArray() {
+        Tag[] result = new Tag[tags.size()];
+        int i = 0;
+        for (Entry<String, String> entry : tags.entrySet()) {
+            result[i] = new Tag(entry.getKey(), entry.getValue());
```

Comment:
I wonder if `Tag` should get a `clone` method instead of using `new Tag(...) ` to make life easier for people.

Reply:
Tags are immutable so it's probably not needed to clone them, you can just pass the references you already have, wherever you like (and you save memory by doing that).  And here we don't have references to `Tag` objects anyway. 


But perhaps it would be a good idea to change the underlying `tags` field to store `List<Tag>` instead of `Map<String, String>` (and update all the `get*` methods accordingly)?? I think we're doing quite stupid stuff now - look at the constructor, it creates a `Tag` instance, passes that to the `addTag` method, which then unrolls it back into two strings and stores them in a map. That is stupid.

Reply:
if we change it to List<Tag> we need to check to make sure each key can only exist once in the list (to not change behaviour from current TreeMap).

Reply:
in general though, I think I'm in favor of using a List to store this rather than a Map.  Maps are slow to add to and retrieve from compared to short lists too.

Reply:
Good point about the uniqueness. Then I'd suggest we use a `HashSet` - that will ensure uniqueness and should not be much slower than a list. I'm not sure but I think we don't need to maintain the order, so a pure `HashSet` should work for us. If we need to maintain the order then we switch to `LinkedHashSet`

Reply:
Hmm I'm kinda having second thoughts about this change.
The implementations will most often want to get the `String` keys/values rather than `Tag`s (for exporting purposes), so this would mean a lot of expensive "unboxing" during export (if we internally store the "boxed" tags). And storing Tags instead of Strings will increase memory consumption as well. 
It's ok that the `MetricID` API can consume `Tag` instances, but providing `String` should be ranked higher than providing `Tag`, IMO.

Reply:
the most frequent case - and therefore the one we want to worry most about for performance - will be responding to a call to /metrics.  While the OpenMetrics format in our spec doesn't say that you need to output tags in any sorted order (interesting it does say it for JSON format) it would seem better IMO if all exporters did sort them.  Assuming we do want them to be sorted by exporters, it would be good to store the tags in a sorted format that is fast to retrieve from.

Reply:
Ok, well I didn't do this change after all so we will still have a sorted `TreeMap<String, String>` that sorts the keys alphabetically. So can this be merged now?